client/src/main/java/io/pravega/client/stream/impl/WatermarkReaderImpl.java

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright (c) 2019 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.stream.impl;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import io.pravega.client.SynchronizerClientFactory;
+import io.pravega.client.segment.impl.Segment;
+import io.pravega.client.state.Revision;
+import io.pravega.client.state.RevisionedStreamClient;
+import io.pravega.client.state.SynchronizerConfig;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.TimeWindow;
+import io.pravega.client.watermark.WatermarkSerializer;
+import io.pravega.common.concurrent.LatestItemSequentialProcessor;
+import io.pravega.shared.segment.StreamSegmentNameUtils;
+import io.pravega.shared.watermarks.Watermark;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.Synchronized;
+
+public class WatermarkReaderImpl {
+
+    private final Stream stream;
+    private final WatermarkFetcher fetcher;
+    private final LatestItemSequentialProcessor<Map<SegmentWithRange, Long>> processor;
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final ArrayDeque<Watermark> inflight = new ArrayDeque<>();
+    @GuardedBy("lock")
+    private Long passedTimestamp = null;
+    
+    @RequiredArgsConstructor
+    private static class WatermarkFetcher {
+        private final RevisionedStreamClient<Watermark> client; 
```

Comment:
the client needs to be closed to release the connection.

Reply:
Done